ibtk/src/solvers/impls/FACPreconditioner.cpp

```diff
@@ -94,48 +96,43 @@ FACPreconditioner::setTimeInterval(const double current_time, const double new_t
 } // setTimeInterval
 
 bool
-FACPreconditioner::solveSystem(SAMRAIVectorReal<NDIM, double>& u, SAMRAIVectorReal<NDIM, double>& f)
+FACPreconditioner::solveSystem(SAMRAIVectorReal<NDIM, double>& x, SAMRAIVectorReal<NDIM, double>& b)
 {
     // Initialize the solver, when necessary.
     const bool deallocate_after_solve = !d_is_initialized;
-    if (deallocate_after_solve) initializeSolverState(u, f);
+    if (deallocate_after_solve) initializeSolverState(x, b);
 
     // Set the initial guess to equal zero.
-    u.setToScalar(0.0, /*interior_only*/ false);
+    x.setToScalar(0.0, /*interior_only*/ false);
+
+    // Clone the right-hand-side vector to avoid modifying it during the
+    // preconditioning operation.
+    ScopedVectorCopy<double> f(b);
+    ScopedVectorDuplicate<double> r(b);
```

Comment:
I agree that we should avoid extra data copies. What's going on here is that we set up `f` as a deep copy of `b` and `r` as a vector of the same size with all zeros. This avoids one extra data copy (we don't need to set `r = f` - that's done in the residual computation)